src/render/src/lib.rs

```diff
@@ -146,70 +140,259 @@ pub mod shade;
+
+        let swap_config = core::SwapchainConfig::new()
+            .with_color::<Cf>();
+        // TODO: call supports_queue
+        let mut swapchain = surface.build_swapchain(swap_config, &queue.inner);
+        let backbuffers = swapchain.take_backbuffers();
+
+        let frame_bundles = backbuffers.iter()
+            .map(|_| FrameBundle {
+                handles: handle::Bag::new(),
+                scopes: Vec::new(),
+                wait_semaphore: Sync::reached(device.create_semaphore()),
+                signal_semaphore: Sync::reached(device.create_semaphore()),
+                signal_fence: Sync::reached(device.create_fence(true)),
+            }).collect();
 
         let (garbage_sender, garbage_receiver) = handle::garbage_channel();
+        let mut device = Device::new(device, heap_types, memory_heaps, garbage_sender);
+
+        let backbuffers = backbuffers.into_iter()
+            .map(|raw| {
+                Backbuffer {
+                    color: Typed::new(device
+                        .view_backbuffer_as_render_target_raw(
+                            raw.color,
+                            unimplemented!(),
+                            unimplemented!(),
+                            Cf::get_format(),
+                            (0..1, 0..1)
+                        ).unwrap()
+                    )
+                }
+            }).collect::<Vec<_>>();
 
-        Gpu {
-            device: Device::new(device, heap_types, memory_heaps, garbage_sender),
+        (Context {
+            surface,
+            device,
             queue,
+            swapchain,
+            frame_bundles,
+            frame_acquired: None,
             garbage: garbage_receiver,
+        }, backbuffers)
+    }
+
+    pub fn acquire_frame(&mut self) -> Frame {
+        assert!(self.frame_acquired.is_none());
+
+        let mut bundle = self.frame_bundles.pop_front().unwrap();
+
+        if bundle.signal_fence.state == Unreached {
+            self.device.mut_raw()
+                .wait_for_fences(
+                    &[&bundle.signal_fence.inner],
+                    core::device::WaitFor::All,
+                    !0);
         }
+        self.device.mut_raw().reset_fences(&[&bundle.signal_fence.inner]);
+        bundle.signal_fence.state = Reached;
+
+        bundle.handles.clear();
+        bundle.scopes.clear();
+
+        let frame = self.swapchain.acquire_frame(
+            core::FrameSync::Semaphore(&mut bundle.wait_semaphore.inner)
+        );
+        bundle.wait_semaphore.state = Unreached;
+        self.frame_acquired = Some(bundle);
+
+        self.cleanup();
+        frame
+    }
+
+    pub fn acquire_graphics_scope(&mut self) -> encoder::Scope<B, core::Graphics> {
+        self.queue.acquire_encoder_scope()
     }
 
-    pub fn cleanup(&mut self) {
-        use core::Device;
+    // TODO: allow submissions before present
+    pub fn present(&mut self, submits: Vec<encoder::Submit<B, core::Graphics>>) {
```

Comment:
Is there something preventing me from submitting command buffers from a pool which I requested in a previous frame?

Reply:
No, that would extend the lifetime of the associated scope (with `ScopeDependency`) and keep it from being reset until unused.

Reply:
The pool gets send back to the queue and added to the current frame once everything has been submitted if I understand it correctly?

Reply:
`Encoder` and `Submit` hold a scope dependency. On submission it is also added to `FrameBundle`. Once a frame fence is waited upon, the associated `FrameBundle` dependencies are cleared. The pool gets send back to the queue when there is no encoder, submit or frame still referencing the scope. Only then it can be reset and recycled into another scope.

Reply:
Okay, I think I grasped everything. Thanks for explanation! Looks solid!